stateresolution.go

```diff
@@ -0,0 +1,255 @@
+//  * A List of resolved events.
+//
+// It implements the AuthEvents interface and can be used for running auth checks.
+type stateResolver struct {
+	creates                   []Event
+	powerLevels               []Event
+	joinRules                 []Event
+	thirdPartyInvites         [][]Event
+	members                   [][]Event
+	others                    [][]Event
+	resolvedCreate            *Event
+	resolvedPowerLevels       *Event
+	resolvedJoinRules         *Event
+	resolvedThirdPartyInvites map[string]*Event
+	resolvedMembers           map[string]*Event
+	result                    []Event
+}
+
+func (r *stateResolver) Create() (*Event, error) {
+	return r.resolvedCreate, nil
+}
+
+func (r *stateResolver) PowerLevels() (*Event, error) {
+	return r.resolvedPowerLevels, nil
+}
+
+func (r *stateResolver) JoinRules() (*Event, error) {
+	return r.resolvedJoinRules, nil
+}
+
+func (r *stateResolver) ThirdPartyInvite(key string) (*Event, error) {
+	return r.resolvedThirdPartyInvites[key], nil
+}
+
+func (r *stateResolver) Member(key string) (*Event, error) {
+	return r.resolvedMembers[key], nil
+}
+
+func (r *stateResolver) addConflicted(events []Event) {
+	type conflictKey struct {
+		eventType string
+		stateKey  string
+	}
+	offsets := map[conflictKey]int{}
+	// Split up the conflicted events into blocks with the same state key.
+	for _, event := range events {
+		key := conflictKey{event.Type(), *event.StateKey()}
+		// Work out which block to add the event to.
+		var block *[]Event
+		// By default we add the event to a block in the others list.
+		blockList := &r.others
+		switch key.eventType {
+		case "m.room.create":
+			if key.stateKey == "" {
+				block = &r.creates
+			}
+		case "m.room.power_levels":
+			if key.stateKey == "" {
+				block = &r.powerLevels
+			}
+		case "m.room.join_rules":
+			if key.stateKey == "" {
+				block = &r.joinRules
+			}
+		case "m.room.member":
+			blockList = &r.members
+		case "m.room.third_party_invite":
+			blockList = &r.thirdPartyInvites
+		}
+		if block == nil {
+			// We need to find an entry for the state key in a block list.
+			offset, ok := offsets[key]
+			if !ok {
+				// This is the first time we've seen that state key so we add a
+				// new block to the block list.
+				offset = len(*blockList)
+				*blockList = append(*blockList, nil)
+			}
+			// Get the address of the block in the block list.
+			block = &(*blockList)[offset]
+		}
+		// Add the event to the block.
+		*block = append(*block, event)
+	}
+}
+
+// Add an event to the resolved auth events.
+func (r *stateResolver) addAuthEvent(event *Event) {
+	switch event.Type() {
+	case "m.room.create":
+		r.resolvedCreate = event
+	case "m.room.power_levels":
+		r.resolvedPowerLevels = event
+	case "m.room.join_rules":
+		r.resolvedJoinRules = event
+	case "m.room.member":
+		r.resolvedMembers[*event.StateKey()] = event
+	case "m.room.third_party_invite":
+		r.resolvedThirdPartyInvites[*event.StateKey()] = event
+	}
```

Comment:
And if the event type isn't any of those? That feels like a logic error and I feel we should panic. We should never encounter that. I'd rather we failed fast rather than silently continued like we do currently.